### PR TITLE
ar71xx: add support for TP-Link CPE210 v2

### PR DESCRIPTION
**Pull request by adschm**
Since PR #642 has been open for a long time, this is an attempt to finalize the support of the CPE210 v2.

I hope the original author will forgive my impatience.

Changes compared to the last state of PR #642:
- Fixed partitions in tplink-safeloader.c
- Reduced duplicate code in target/linux/ar71xx/base-files/lib/ar71xx.sh
- Reduced duplicate code in target/linux/ar71xx/base-files/lib/upgrade/platform.sh

If my "duplicate" PR is against protocol, please tell me and/or close it.

### CODE REVIEW

**Comment by robimarko** [comment]
@adrianschmutzler No issues,I will close mine as I have exams coming up and various projects at work.

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh [hunk 1]
```diff
@@ -575,7 +578,11 @@ platform_check_image() {
 	eap120|\
 	wbs210|\
 	wbs510)
-		tplink_pharos_check_image "$1" && return 0
+		tplink_pharos_check_image "$1" "7f454c46" "$(tplink_pharos_get_model_string)" '' && return 0
+		return 1
+		;;
+	cpe210-v2)
+		tplink_pharos_v2_check_image "$1" "01000000" "$(tplink_pharos_v2_get_model_string)" '\0\xff\r' && return 0
```
Comment (genofire) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: Sry my stupped question - where is `tplink_pharos_v2_check_image` could only find `tplink_pharos_check_image`

**Review by genofire — COMMENTED** [review]
A question

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh (hunk 1 above)
Comment (adrianschmutzler) [reply target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: Good find. I overlooked this when merging tplink_pharos_v2_check_image and tplink_pharos_check_image.
Commit should already be updated!

**Comment by achterin** [comment]
Is there any reason why this pr didn't get merged already? I can confirm everything works as expected.

**Comment by rotanid** [comment]
@achterin i guess the reason is, that no volunteer had the time and/or interest so far to check it...

**Comment by adrianschmutzler** [comment]
I think this commit breaks the PR:

https://github.com/openwrt/openwrt/commit/805e59cdcc2b6b2bb3f7583d8ccdb932e6cdeb7b

Any ideas on how to implement it? Or should we just keep the old loader and stop inheriting from the cpe210v1 config?

**Comment by rotanid** [comment]
@adrianschmutzler maybe the author of that commit can answer this best: @NeoRaider

**Comment by NeoRaider** [comment]
Either way is fine. The okli-loader makes the flash layout more flexible (as described in the commit message of 805e59c), but as the flash layout should be fine for some kernel growth with the current 1.75MB os-image, you can also use the old image build commands. If you go with the old commands, you can also try removing `patch-cmdline` and switching from `loader-kernel` to `loader-kernel-cmdline` instead - my lzma-loader fixes *might* have fixed `loader-kernel-cmdline` for the ELF loader type (`loader-kernel-cmdline` was broken in this setup before, so we resorted to using `patch-cmdline` to get the cmdline into the image.)

I'm a bit surprised that it doesn't just work with okli-loader without further changes, as CPE210 v1 and v2 have the same offset to the os-image partition...

**Comment by adrianschmutzler** [comment]
I didn't test it, so it may still work.
I just was stirred up by the additional definitions
153   +  LOADER_FLASH_OFFS := 0x43000  
 154   +  COMPILE := loader-$(1).elf  
 155   +  COMPILE/loader-$(1).elf := loader-okli-compile  
since they are inherited and I do not know whether they will do "damage" if I do not overwrite remove them (while keeping the current KERNEL definition in v2!).

Your change won't change the CPE210v2 to okli-leader, since the definition for v2 has its own KERNEL := ... line and thus this part is not affected by the commit 805e59c.

Since I don't feel capable of really implementing/testing the new loader, either someone else would have to help me here or I will stick to old one. For the old one, I would tend to not inherit from Device/cpe510-520-v1 anymore, as this gave me problems more than once.

Opinions?

**Comment by NeoRaider** [comment]
Ah, I overlooked the KERNEL definition. In this case, I'd just recommend not inheriting the v1 definitions anymore. Common settings like mtdparts could be moved into a new define inherited by both cpe510-520-v1 and cpe210-v1.

**Comment by adrianschmutzler** [comment]
Fixed as suggested and rebased.
Did not have time to test it on device ...

**Comment by NeoRaider** [comment]
I've verified that CPE510 v1 is working fine with these changes, including sysupgrade. It would be great if you could give your last changes a final test run on the CPE210 v2, then we can get this merged.

**Comment by adrianschmutzler** [comment]
Run Test successful. Only tested with sysupgrade...

**Comment by NeoRaider** [comment]
Thanks, merged!

I noticed 2 minor issues when merging, which I fixed up:
* Multiple Signed-off-by lines are usually added in the order that people worked on the code (as it is also done by `git commit -s`)
* I moved the new Device/cpexxx section from the top to the actual cpe device definitions

**Comment by adrianschmutzler** [comment]
@NeoRaider 
There are code guidelines which said that the Author and Commit field should be the same person. However, GitHub only considers the first commit name here, so this is why I did it this way.
From a logical point of view, I also would have chosen "your" order.

Thanks for merging.

**Pull request closed without merging**